client/k8s.go

```diff
@@ -164,80 +133,135 @@ func homeDir() string {
 
 	if err != nil {
-		klog.Errorf("Error while while getting storage class: %s\n", err)
-		os.Exit(1)
+		return nil, errors.Wrap(err, "error while while getting storage class")
 	}
 
-	return scs
+	return scs, nil
 }
 
 // GetCSIVolume using the K8sClient's storage class client
-func (k K8sClient) GetCSIVolume(volname string) *v1.VolumeAttachment {
-	vol, err := k.sc.VolumeAttachments().Get(volname, metav1.GetOptions{})
+func (k K8sClient) GetCSIVolume(volname string) (*v1.VolumeAttachment, error) {
+	vol, err := k.K8sCS.StorageV1().VolumeAttachments().Get(volname, metav1.GetOptions{})
 
 	if err != nil {
-		klog.Errorf("Error while while getting volumes: %s\n", err)
-		os.Exit(1)
+		return nil, errors.Wrap(err, "error while while getting storage csi volume")
 	}
 
-	return vol
+	return vol, nil
 }
 
 // GetcStorVolumes using the K8sClient's storage class client
-func (k K8sClient) GetcStorVolumes() cstorv1.CStorVolumeList {
+func (k K8sClient) GetcStorVolumes() (*cstorv1.CStorVolumeList, error) {
 
-	cStorVols, err := k.cStorCS.CStorVolumes("").List(metav1.ListOptions{})
+	cStorVols, err := k.OpenebsCS.CstorV1().CStorVolumes("").List(metav1.ListOptions{})
 
 	if err != nil {
-		klog.Errorf("Error while while getting volumes: %s\n", err)
-		os.Exit(1)
+		return nil, errors.Wrapf(err, "Error while while getting volumes")
 	}
 
-	return *cStorVols
+	return cStorVols, nil
 
 }
 
 // GetcStorVolume fetches the volume object of the given name in the given namespace
-func (k K8sClient) GetcStorVolume(volName string, namespace string) *cstorv1.CStorVolume {
-	vols := k.cStorCS.CStorVolumes(namespace)
+func (k K8sClient) GetcStorVolume(volName string) (*cstorv1.CStorVolume, error) {
+	vols := k.OpenebsCS.CstorV1().CStorVolumes(k.ns)
 	volInfo, err := vols.Get(volName, metav1.GetOptions{})
 
 	if err != nil {
-		klog.Errorf("Error while while getting volume %s in %s namespace %s\n",
-			volName, namespace, err)
-		os.Exit(1)
+		return nil, errors.Wrapf(err, "error while while getting volume %s", volName)
 	}
 
-	return volInfo
+	return volInfo, nil
 }
 
-// GetcStorPVCs used to get the infor for the underlying
+// GetcStorPVCs used to get the info for for the underlying
 // PVC
-func (k K8sClient) GetcStorPVCs(node string) map[string]*util.Volume {
+func (k K8sClient) GetcStorPVCs(node string) (map[string]*util.Volume, error) {
 
 	volumes := make(map[string]*util.Volume)
 
-	PVCs, err := k.sc.VolumeAttachments().List(metav1.ListOptions{})
+	PVCs, err := k.K8sCS.StorageV1().VolumeAttachments().List(metav1.ListOptions{})
 	if err != nil {
-		klog.Errorf("Error while while getting storage volume attachments on %s", node, err)
-		os.Exit(1)
+		return volumes, errors.Wrap(err, "error while while getting storage volume attachments")
 	}
 
 	for _, i := range PVCs.Items {
+		pv, err := k.GetPV(*i.Spec.Source.PersistentVolumeName)
+		if err != nil {
+			klog.Errorf("Failed to get PV", i.ObjectMeta.Name)
```

Comment:
Can we add `continue` after logging the error? Because if some problem occurs while fetching the corresponding PV then the value will be `nil` and accessing over nil `pv.Spec.AccessModes` will crash the process.

Reply:
i tried to deal with this in the check and above loop now, do you think this is ok ? did it this way because we could still print other details in case this field was missing, if you don't think this is ok will just do it the way you have asked me to . :)

Reply:
Still `continue` over here required think a case where `i.Spec.Source.PersistentVolumeName` is not nil and while fetching PV from etcd using `GetPV()` caused n/w error and you will get nil over here because `GetPV()` is returning nil. So to handle such scenario need to add `continue`

Reply:
AccessModeToString is handling that nil ... right ?